### PR TITLE
Generate a runtime crash when assigning scalar to unallocated array

### PR DESCRIPTION
**Pull request by jeanPerier**

### CODE REVIEW

**Review by psteinfeld — APPROVED** [review]
All builds, tests, is formatted correctly and looks good.

File: flang/lib/Optimizer/Builder/Runtime/Stop.cpp [hunk 1]
```diff
@@ -20,3 +21,18 @@ void fir::runtime::genExit(fir::FirOpBuilder &builder, mlir::Location loc,
       fir::runtime::createArguments(builder, loc, exitFunc.getType(), status);
   builder.create<fir::CallOp>(loc, exitFunc, args);
 }
+
+void fir::runtime::genCrash(fir::FirOpBuilder &builder, mlir::Location loc,
```
Comment (vdonaldson) [inline flang/lib/Optimizer/Builder/Runtime/Stop.cpp:25]: Would it help to have a name that includes "message", like `genCrashMessage`?

**Review by vdonaldson — APPROVED** [review]

File: flang/include/flang/Optimizer/Builder/Runtime/Stop.h [hunk 2]
```diff
@@ -23,5 +24,9 @@ namespace fir::runtime {
 /// Generate call to EXIT intrinsic runtime routine.
 void genExit(fir::FirOpBuilder &, mlir::Location, mlir::Value status);
 
+/// Generate call to crash the program with an error message when detecting
+/// an invalid situation at runtime.
+void genCrash(fir::FirOpBuilder &, mlir::Location, const std::string &message);
```
Comment (schweitzpgi) [inline flang/include/flang/Optimizer/Builder/Runtime/Stop.h:29]: Should use `llvm::StringRef` in preference to `const std::string &`.

File: flang/lib/Optimizer/Builder/Runtime/Stop.cpp (hunk 1 above)
Comment (sscalpone) [reply flang/lib/Optimizer/Builder/Runtime/Stop.cpp:25]: It's not really a crash.  It's a user error.

genReportUserError() ?

@psteinfeld The name in the runtime -- Crash() -- will show up in the user's traceback.  Maybe we could change it too?

Comment (jeanPerier) [reply flang/lib/Optimizer/Builder/Runtime/Stop.cpp]: Thanks @sscalpone, your point makes sense to me. I would just keep a `Fatal` somewhere to make it clear the program will not continue. Maybe genReportFatalUserError  (I have updated this PR to that, but not the runtime) ?

**Comment by psteinfeld** [comment]
Thanks, Steve.

I’ll implement this.

Pete

File: flang/lib/Optimizer/Builder/Runtime/Stop.cpp (hunk 1 above)
Comment (psteinfeld) [reply flang/lib/Optimizer/Builder/Runtime/Stop.cpp]: See https://reviews.llvm.org/D118703.

**Review by schweitzpgi — APPROVED** [review]

**Comment by jeanPerier** [comment]
> I’ll implement this.
Pete

Thanks, I cherry picked your commit from LLVM here.

**Comment by psteinfeld** [comment]
> Thanks, I cherry picked your commit from LLVM here.
Thanks, Jean.  You're the best.

**Comment by sscalpone** [comment]
@psteinfeld More exciting read here:  https://docs.nvidia.com/hpc-sdk/compilers/hpc-compilers-user-guide/#env-vars-nv-term

**Pull request merged**